cylc/flow/etc/job.sh

```diff
@@ -140,6 +135,34 @@ cylc__job__main() {
     exit 0
 }
 
+###############################################################################
+# Run user scripts in subshell to protect cylc job script from interference.
+# Getting signal traps to run immediately requires waiting on a background
+# process inside a function (not sure why function is needed for this, but
+# without it tests/job-file-trap/00-sigusr1.t fails).
+cylc__job__run_user_scripts() {
+    typeset func_name=
+    for func_name in 'env_script' 'user_env' 'pre_script' 'script' 'post_script'; do
+        cylc__job__run_inst_func "${func_name}"
+    done &
+    CYLC_TASK_SCRIPT_PID=$!
+    wait "${CYLC_TASK_SCRIPT_PID}"
+}
```

Comment:
This essentially implements the suggested solution from #2043. Using background subshell and `wait` allows the trap to execute immediately after receiving the signal rather than after the execution of current command completes.